papers/ring_vrf/zk_cont.tex

```diff
@@ -1,34 +1,101 @@
-as well as this additional proof-of-knowledge.
-Zero-knowledge however should hold even if \Reprove gets invoked multiple
-times upon the same \Preprove results,
- again even with the additional proof-of-knowledge.
+\item $\ZKCont.\Setup: (1^{\lambda}) \mapsto \crs$ a setup algorithm that on input the security parameter 
+outputs a common reference string $\crs$, 
+
+\item $\ZKCont.\Gen: (\crs, \relone) \mapsto (\crspk, \crsvk)$ \, 
+outputs a pair of proving key $\crspk$ and verification key $\crsvk$, 
+
+\item $\ZKCont.\Preprove: (\crspk, \bar{y}, \bar{x}, \bar{\omega}, \relone) \mapsto (X, \pi, b)$ \,
+constructs commitment $X$ from a vector of inputs $\bar{x}$ (called \emph{opaque}) and 
+constructs proof $\pi$ from vector of inputs 
+$\bar{y}$ (called \emph{transparent}), from $\bar{x}$ and vector of witnesses $\bar{\omega}$, and 
+also outputs $b$ as the opening for $X$,
+
+\item $\ZKCont.\Reprove: (\crspk, X, \pi, \relone) \mapsto (X', \pi', b')$ \,
+ finalises commitment $X'$ and proof $\pi'$ and returns an opening $b'$ for the commitment, 
+ 
+\item $\ZKCont.\Verify: (\crsvk, \bar{y}, X', \pi', \relone) \mapsto 0/1$ \, outputs $1$ in case it accepts and $0$ otherwise,
+\end{itemize}
+and satisfies perfect completeness for $\Preprove$ and for $\Reprove$,  knowledge soundness and zero-knowledge as defined below:\\
+%TO DO: Re-write this: We define (white-box) knowledge soundness for zero-knowledge continuations
+%exactly like for zero-knowledge proofs, but with the composition 
+%$\Prove : (\bar{y}, \bar{x}; \bar{\omega}) \mapsto \Reprove(\Preprove(\bar{y}, \bar{x}; \bar{\omega}))[0]$
+%as well as this additional proof-of-knowledge.
+%Zero-knowledge however should hold even if \Reprove gets invoked multiple
+%times upon the same \Preprove results,
+%again even with the additional proof-of-knowledge.
+\noindent \textbf{Perfect Completeness for $\Preprove$} For every $(\bary, \barx; \baromega) \in \relone$ it holds:
+\begin{align*}
+\mathit{Pr} (& \ZKCont.\Verify(\crsvk, \bar{y}, X, \pi, \relone) = 1 \ | \ \\ 
+                   & \crs \leftarrow \ZKCont.\Setup (1^{\lambda}), (\crspk, \crsvk) \leftarrow \ZKCont.\Gen(\crs, \relone), \\ 
+                   & (X, \pi, b) \leftarrow \ZKCont.\Preprove(\crspk, \bar{y}, \bar{x}, \bar{\omega}, \relone)) = 1
+\end{align*}
+
+\noindent \textbf{Perfect Completeness for $\Reprove$} For every efficient adversary $A$ it holds: 
+\begin{align*}
+\mathit{Pr} (& \ZKCont.\Verify(\crsvk, \bar{y}, X, \pi, \relone) = 1  = >  \ZKCont.\Verify(\crsvk, \bar{y}, X', \pi', \relone) = 1  \ | \ \\ 
+                   & \crs \leftarrow \ZKCont.\Setup (1^{\lambda}), (\crspk, \crsvk) \leftarrow \ZKCont.\Gen(\crs, \relone), \\ 
+                   & (\bary, X, \pi) \leftarrow A(\crs,\relone), (X', \pi', b') \leftarrow \ZKCont.\Reprove(\crspk, X, \pi, \relone)) = 1
+\end{align*}
+\noindent \textbf{Knowledge Soundness} For every efficient adversary $A$ there exists an efficient extractor $E$ such that 
```

Comment:
Something is strange here because it seems that for every adversary that outputs (\bary, \barx, b, \pi), there is an extractor. I think it should be for every adv that that outputs (\bary, \barx, b, \pi) where \pi is verified. But I am not sure how you can define verified pi here because A does not output X which is part of the Verify algorithm.

Reply:
The verifier needs to compute X from  \barx and b and then verify. That's what this definition does with its \ZKCont.\Preprove_{|X}}.

Reply:
I think we should say that \pi verifies in the given part of the probability. Otherwise, it says that for every adv that outputs any (\bary, \barx, b, \pi), we can construct an extractor which outputs a witness.